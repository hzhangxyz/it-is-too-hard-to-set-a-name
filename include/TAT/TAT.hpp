/**
 * \file TAT.hpp
 *
 * Copyright (C) 2019-2021 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#pragma once
#ifndef TAT_HPP
#define TAT_HPP

#ifndef __cplusplus
#error only work for c++
#endif

#ifdef _MSVC_LANG
#if _MSVC_LANG < 201703L
#error require c++17 or later
#endif
#else
#if __cplusplus < 201703L
#error require c++17 or later
#endif
#endif

// 开关说明
// TAT_USE_MPI 定义以开启MPI支持, cmake可对此进行定义
// TAT_USE_MKL_TRANSPOSE 定义以使用mkl加速转置, cmake可对此进行定义 TODO 进一步优化
// TAT_USE_MKL_GEMM_BATCH 定义以使用mkl的?gemm_batch, cmake可对此进行定义
// TAT_USE_SINGULAR_MATRIX svd出来的奇异值使用矩阵表示
// TAT_USE_SIMPLE_NAME 定义以使用原始字符串作为name
// TAT_USE_VALID_DEFAULT_TENSOR 默认tensor初始化会产生一个合法的tensor, 默认不合法
// TAT_USE_TIMER 对常见操作进行计时
// TAT_USE_RESTRICT_SMALL_ALLOCATOR 使用严格的小分配器
// TAT_SMALL_ALLOCATOR_SIZE 小分配器的大小
// TAT_ERROR_BITS 将各类警告转换为异常
// TAT_NOTHING_BITS 将各类警告转换为静默
// TAT_L3_CACHE, TAT_L2_CACHE, TAT_L1_CACHE 在转置中会使用
// TAT_USE_L3_CACHE 转置中默认不使用l3_cache, 设置以使用之

/**
 * TAT is A Tensor library
 */
namespace TAT {
   /**
    * \defgroup Miscellaneous
    * @{
    */

   /**
    * TAT的版本号
    */
   inline const char* version =
#ifdef TAT_VERSION
         TAT_VERSION
#else
         "0.2.0"
#endif
         ;

   /**
    * 编译与license的相关信息
    */
   inline const char* information = "TAT"
#ifdef TAT_VERSION
                                    " " TAT_VERSION
#endif
                                    " ("
#ifdef TAT_BUILD_TYPE
                                    "" TAT_BUILD_TYPE ", "
#endif
                                    "" __DATE__ ", " __TIME__
#ifdef TAT_COMPILER_INFORMATION
                                    ", " TAT_COMPILER_INFORMATION
#endif
                                    ")\n"
                                    "Copyright (C) 2019-2021 Hao Zhang<zh970205@mail.ustc.edu.cn>\n"
                                    "This is free software; see the source for copying conditions.  There is NO\n"
                                    "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.";

   /// \private
   struct evil_t {
      evil_t() noexcept;
      ~evil_t();
   };
   /**
    * Debug模式中, 将在程序末尾打印一行友情提示, 过早的优化是万恶之源, 同时此对象也控制windows下终端的色彩模式
    */
   inline const evil_t evil;

   /**
    * TAT使用的日志打印
    */
   inline void TAT_log(const char* message);

   /**
    * 什么事情也不做
    *
    * 接口和`TAT_warning`, `TAT_error`一致, 可供各种细分的警告或错误使用, 通过设置他们为这三个中的一个的指针来选择对错误的容忍程度
    *
    * \see TAT_warning, TAT_error
    */
   inline void TAT_nothing(const char*) {}

   /**
    * TAT使用的打印警告
    *
    * \param message 待打印的内容
    *
    * \see TAT_nothing, TAT_error
    */
   inline void TAT_warning(const char* message);

   /**
    * TAT使用的抛出运行时异常
    *
    * \param message 异常说明
    *
    * \see TAT_nothing, TAT_warning
    */
   inline void TAT_error(const char* message);

   // 下面这一段使用两个BITS来选择一些情况下是否警告，是否静默或者是否直接报错退出

#ifndef TAT_DOXYGEN_SHOULD_SKIP_THIS

#ifndef TAT_ERROR_BITS
#define TAT_ERROR_BITS 0
#endif
#ifndef TAT_NOTHING_BITS
#define TAT_NOTHING_BITS 0
#endif

   constexpr auto TAT_warning_or_error_when_lapack_error = TAT_ERROR_BITS & 1 ? TAT_error : TAT_NOTHING_BITS & 1 ? TAT_nothing : TAT_warning;
   constexpr auto TAT_warning_or_error_when_name_missing = TAT_ERROR_BITS & 2 ? TAT_error : TAT_NOTHING_BITS & 2 ? TAT_nothing : TAT_warning;
   constexpr auto TAT_warning_or_error_when_copy_shared = TAT_ERROR_BITS & 4 ? TAT_error : TAT_NOTHING_BITS & 4 ? TAT_nothing : TAT_warning;
#endif
   /**@}*/
} // namespace TAT

#include <cstdint>

namespace TAT {
   // 下面三个类型原本是short, int, long
   // 在linux(lp64)下分别是16, 32, 64
   // 但是windows(llp64)中是16, 32, 32
   // 在io中会出现windows和linux的输入输出互相不可读的问题
   // 所以显式写成uintxx_t的格式
   // TAT中还有一些地方会出现int, 一般为调用blas和lapack的地方
   // 为32位, 在64位系统中(ilp64)和32位系统中(lp32)分别是64为和16位
   // 所以底层的blas和lapack库不可以是ilp64或者lp32版本
   /**
    * 张量的秩的类型
    */
   using Rank = std::uint16_t;
   /**
    * 张量分块数目和一个边上对称性数目的类型
    */
   using Nums = std::uint32_t;
   /**
    * 张量数据维度大小和数据本身大小的类型
    */
   using Size = std::uint64_t;

   /**
    * 费米箭头方向的类型, `false`和`true`分别表示出入
    */
   using Arrow = bool;
} // namespace TAT

#include <complex>
#include <type_traits>

namespace TAT {
   template<typename T>
   concept is_real = std::is_scalar_v<T>;

   template<typename T>
   concept is_complex = std::is_same_v<T, std::complex<typename T::value_type>>;

   template<typename T>
   concept is_scalar = is_real<T> || is_complex<T>;

   template<typename T>
   struct real_scalar_helper : std::conditional<std::is_scalar_v<T>, T, void> {};
   template<typename T>
   struct real_scalar_helper<std::complex<T>> : std::conditional<std::is_scalar_v<T>, T, void> {};

   /**
    * 取对应的实数类型, 在svd, norm等地方会用到
    * \tparam T 如果`T`是`std::complex<S>`, 则为`S`, 若`T`为其他标量类型, 则为`T`本身, 否则为`void`
    */
   template<typename T>
   using real_scalar = typename real_scalar_helper<T>::type;
} // namespace TAT

#include "structure/tensor.hpp"

#include "miscellaneous/io.hpp"
#include "miscellaneous/mpi.hpp"
#include "miscellaneous/scalar.hpp"

#include "implement/contract.hpp"
#include "implement/edge_miscellaneous.hpp"
#include "implement/edge_operator.hpp"
#include "implement/exponential.hpp"
#include "implement/get_item.hpp"
#include "implement/identity.hpp"
#include "implement/multiple.hpp"
#include "implement/qr.hpp"
#include "implement/shrink_and_expand.hpp"
#include "implement/svd.hpp"
#include "implement/trace.hpp"
#include "implement/transpose.hpp"

namespace TAT {
   /**
    * Z2对称性的类型
    */
   using Z2 = bool;
   /**
    * U1对称性的类型
    */
   using U1 = std::int32_t;

   using NoSymmetry = Symmetry<>;
   using Z2Symmetry = Symmetry<Z2>;
   using U1Symmetry = Symmetry<U1>;
   using FermiSymmetry = Symmetry<fermi_wrap<U1>>;
   using FermiZ2Symmetry = Symmetry<fermi_wrap<U1>, Z2>;
   using FermiU1Symmetry = Symmetry<fermi_wrap<U1>, U1>;
} // namespace TAT

#endif
