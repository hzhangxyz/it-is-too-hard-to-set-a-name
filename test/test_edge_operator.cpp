/**
 * Copyright (C) 2019-2021 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <TAT/TAT.hpp>

#include "run_test.hpp"

void run_test() {
   std::cout << TAT::Tensor<double, TAT::NoSymmetry>{{"A", "B"}, {8, 8}}.range() << "\n";
   std::cout << TAT::Tensor<double, TAT::NoSymmetry>{{"A", "B"}, {8, 8}}
                      .range()
                      .edge_rename({{"A", "C"}})
                      .edge_operator(
                            {{"C", {{"D", 4}, {"E", 2}}}, {"B", {{"F", 2}, {"G", 4}}}},
                            {"D", "F"},
                            {{"I", {"D", "F"}}, {"J", {"G", "E"}}},
                            {"J", "I"})
             << "\n";
   std::cout << TAT::Tensor<>{{"A", "B", "C"}, {2, 3, 4}}.range().edge_operator({}, {}, {}, {"B", "C", "A"}) << '\n';
   do {
      auto a =
            TAT::Tensor<double, TAT::U1Symmetry>{
                  {"Left", "Right", "Up", "Down"},
                  {{{-1, 3}, {0, 1}, {1, 2}}, {{-1, 1}, {0, 4}, {1, 2}}, {{-1, 2}, {0, 3}, {1, 1}}, {{-1, 1}, {0, 3}, {1, 2}}}}
                  .set([]() {
                     static double i = 0;
                     return i += 1;
                  });
      auto b = a.edge_rename({{"Right", "Right1"}}).split_edge({{"Down", {{"Down1", {{{0, 1}, {1, 2}}}}, {"Down2", {{{-1, 1}, {0, 1}}}}}}});
      auto c = b.transpose({"Down1", "Right1", "Up", "Left", "Down2"});
      auto d = c.merge_edge({{"Left", {"Left", "Down2"}}});
      auto total = a.edge_rename({{"Right", "Right1"}})
                         .edge_operator(
                               {{"Down", {{"Down1", {{{0, 1}, {1, 2}}}}, {"Down2", {{{-1, 1}, {0, 1}}}}}}},
                               {},
                               {{"Left", {"Left", "Down2"}}},
                               {"Down1", "Right1", "Up", "Left"});
      std::cout << (total - d).norm<-1>() << "\n";
   } while (false);
   do {
      auto a =
            TAT::Tensor<double, TAT::FermiSymmetry>{
                  {"Left", "Right", "Up", "Down"},
                  {{{-1, 3}, {0, 1}, {1, 2}}, {{-1, 1}, {0, 4}, {1, 2}}, {{-1, 2}, {0, 3}, {1, 1}}, {{-1, 1}, {0, 3}, {1, 2}}}}
                  .set([]() {
                     static double i = 0;
                     return i += 1;
                  });
      auto b = a.edge_rename({{"Right", "Right1"}}).split_edge({{"Down", {{"Down1", {{{0, 1}, {1, 2}}}}, {"Down2", {{{-1, 1}, {0, 1}}}}}}});
      auto r = b.reverse_edge({"Left"});
      auto c = r.transpose({"Down1", "Right1", "Up", "Left", "Down2"});
      auto d = c.merge_edge({{"Left", {"Left", "Down2"}}});
      auto total = a.edge_rename({{"Right", "Right1"}})
                         .edge_operator(
                               {{"Down", {{"Down1", {{{0, 1}, {1, 2}}}}, {"Down2", {{{-1, 1}, {0, 1}}}}}}},
                               {"Left"},
                               {{"Left", {"Left", "Down2"}}},
                               {"Down1", "Right1", "Up", "Left"});
      std::cout << (total - d).norm<-1>() << "\n";
      std::cout << total << "\n";
   } while (false);
}
