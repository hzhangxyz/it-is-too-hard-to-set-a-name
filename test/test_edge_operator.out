{[32mnames[0m:[A,B],[32medges[0m:[8,8],[32mblocks[0m:[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63]}
{[32mnames[0m:[J,I],[32medges[0m:[8,8],[32mblocks[0m:[0,4,16,20,32,36,48,52,8,12,24,28,40,44,56,60,1,5,17,21,33,37,49,53,9,13,25,29,41,45,57,61,2,6,18,22,34,38,50,54,10,14,26,30,42,46,58,62,3,7,19,23,35,39,51,55,11,15,27,31,43,47,59,63]}
{[32mnames[0m:[B,C,A],[32medges[0m:[3,4,2],[32mblocks[0m:[0,12,1,13,2,14,3,15,4,16,5,17,6,18,7,19,8,20,9,21,10,22,11,23]}
0
0
{[32mnames[0m:[Down1,Right1,Up,Left],[32medges[0m:[{arrow:0,segment:{0:1,1:2}},{arrow:0,segment:{-1:1,0:4,1:2}},{arrow:0,segment:{-1:2,0:3,1:1}},{arrow:1,segment:{-2:3,-1:4,0:3,1:2}}],[32mblocks[0m:{[33m[0,-1,0,1][0m:[-290,-299,-293,-302,-296,-305],[33m[0,-1,1,0][0m:[205,308,309],[33m[0,0,-1,1][0m:[-310,-334,-313,-337,-316,-340,-319,-343,-322,-346,-325,-349,-328,-352,-331,-355],[33m[0,0,0,0][0m:[224,358,370,227,359,371,230,360,372,233,361,373,236,362,374,239,363,375,242,364,376,245,365,377,248,366,378,251,367,379,254,368,380,257,369,381],[33m[0,0,1,-1][0m:[-79,-91,-103,260,-82,-94,-106,261,-85,-97,-109,262,-88,-100,-112,263],[33m[0,1,-1,0][0m:[264,382,386,267,383,387,270,384,388,273,385,389],[33m[0,1,0,-1][0m:[-139,-157,-175,276,-142,-160,-178,277,-145,-163,-181,278,-148,-166,-184,279,-151,-169,-187,280,-154,-172,-190,281],[33m[0,1,1,-2][0m:[193,195,197,194,196,198],[33m[1,-1,-1,1][0m:[-282,-286,-284,-288,-283,-287,-285,-289],[33m[1,-1,0,0][0m:[-199,-291,-300,-201,-294,-303,-203,-297,-306,-200,-292,-301,-202,-295,-304,-204,-298,-307],[33m[1,-1,1,-1][0m:[-1,-3,-5,206,-2,-4,-6,207],[33m[1,0,-1,0][0m:[-208,-311,-335,-210,-314,-338,-212,-317,-341,-214,-320,-344,-216,-323,-347,-218,-326,-350,-220,-329,-353,-222,-332,-356,-209,-312,-336,-211,-315,-339,-213,-318,-342,-215,-321,-345,-217,-324,-348,-219,-327,-351,-221,-330,-354,-223,-333,-357],[33m[1,0,0,-1][0m:[-7,-31,-55,225,-9,-33,-57,228,-11,-35,-59,231,-13,-37,-61,234,-15,-39,-63,237,-17,-41,-65,240,-19,-43,-67,243,-21,-45,-69,246,-23,-47,-71,249,-25,-49,-73,252,-27,-51,-75,255,-29,-53,-77,258,-8,-32,-56,226,-10,-34,-58,229,-12,-36,-60,232,-14,-38,-62,235,-16,-40,-64,238,-18,-42,-66,241,-20,-44,-68,244,-22,-46,-70,247,-24,-48,-72,250,-26,-50,-74,253,-28,-52,-76,256,-30,-54,-78,259],[33m[1,0,1,-2][0m:[-80,-92,-104,-83,-95,-107,-86,-98,-110,-89,-101,-113,-81,-93,-105,-84,-96,-108,-87,-99,-111,-90,-102,-114],[33m[1,1,-1,-1][0m:[-115,-123,-131,265,-117,-125,-133,268,-119,-127,-135,271,-121,-129,-137,274,-116,-124,-132,266,-118,-126,-134,269,-120,-128,-136,272,-122,-130,-138,275],[33m[1,1,0,-2][0m:[-140,-158,-176,-143,-161,-179,-146,-164,-182,-149,-167,-185,-152,-170,-188,-155,-173,-191,-141,-159,-177,-144,-162,-180,-147,-165,-183,-150,-168,-186,-153,-171,-189,-156,-174,-192]}}
